File: testers/sqlserver/kotlin/src/kotlin/testdb/TestInsertTest.kt
package testdb

import org.junit.Assert.*
import org.junit.Test
import java.math.BigDecimal
import java.util.Random

/**
 * Tests for the TestInsert helper class that generates random test data.
 */
class TestInsertTest {
    private val testInsert = TestInsert(Random(42))

    @Test
    fun testCustomersInsert() {
        SqlServerTestHelper.run { c ->
            val row = testInsert.Customers(
                name = "Test Customer",
                email = "test@example.com",
                c = c
            )

            assertNotNull(row)
            assertNotNull(row.customerId)
            assertEquals("Test Customer", row.name)
            assertEquals("test@example.com", row.email)
        }
    }

    @Test
    fun testProductsInsert() {
        SqlServerTestHelper.run { c ->
            val row = testInsert.Products(
                name = "Test Product",
                price = BigDecimal("99.99"),
                c = c
            )

            assertNotNull(row)
            assertNotNull(row.productId)
            assertEquals("Test Product", row.name)
            assertEquals(0, BigDecimal("99.99").compareTo(row.price))
        }
    }

    @Test
    fun testAllScalarTypesInsert() {
        SqlServerTestHelper.run { c ->
            val row = testInsert.AllScalarTypes(c = c)

            assertNotNull(row)
            assertNotNull(row.id)
            assertNotNull(row.colRowversion)
        }
    }

    @Test
    fun testOrdersWithCustomerFK() {
        SqlServerTestHelper.run { c ->
            val customer = testInsert.Customers(
                name = "FK Customer",
                email = "fk@example.com",
                c = c
            )

            val order = testInsert.Orders(
                customerId = customer.customerId,
                c = c
            )

            assertNotNull(order)
            assertEquals(customer.customerId, order.customerId)
        }
    }

    @Test
    fun testOrderItemsWithFKs() {
        SqlServerTestHelper.run { c ->
            val customer = testInsert.Customers(
                name = "OrderItem Customer",
                email = "orderitem@example.com",
                c = c
            )
            val product = testInsert.Products(
                name = "OrderItem Product",
                price = BigDecimal("50.00"),
                c = c
            )
            val order = testInsert.Orders(
                customerId = customer.customerId,
                c = c
            )

            val orderItem = testInsert.OrderItems(
                orderId = order.orderId,
                productId = product.productId,
                c = c
            )

            assertNotNull(orderItem)
            assertEquals(order.orderId, orderItem.orderId)
            assertEquals(product.productId, orderItem.productId)
        }
    }

    @Test
    fun testMultipleInserts() {
        SqlServerTestHelper.run { c ->
            val row1 = testInsert.Customers(name = "Customer1", email = "c1@test.com", c = c)
            val row2 = testInsert.Customers(name = "Customer2", email = "c2@test.com", c = c)
            val row3 = testInsert.Customers(name = "Customer3", email = "c3@test.com", c = c)

            assertNotEquals(row1.customerId, row2.customerId)
            assertNotEquals(row2.customerId, row3.customerId)
            assertNotEquals(row1.customerId, row3.customerId)
        }
    }

    @Test
    fun testInsertWithDifferentValues() {
        SqlServerTestHelper.run { c ->
            val row1 = testInsert.Customers(name = "Alpha", email = "alpha@test.com", c = c)
            val row2 = testInsert.Customers(name = "Beta", email = "beta@test.com", c = c)

            assertEquals("Alpha", row1.name)
            assertEquals("Beta", row2.name)
            assertEquals("alpha@test.com", row1.email)
            assertEquals("beta@test.com", row2.email)
        }
    }
}

File: testers/duckdb/scala/src/scala/testdb/TestInsertTest.scala
package testdb

import org.junit.Assert._
import org.junit.Test

import scala.util.Random

/** Tests for TestInsert functionality - automatic random data generation for testing.
  */
class TestInsertTest {
  private val testInsert = TestInsert(Random(42))

  @Test
  def testCustomersInsert(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val row = testInsert.Customers()
    assertNotNull(row)
    assertNotNull(row.customerId)
    assertNotNull(row.name)
  }

  @Test
  def testCustomersWithCustomization(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val row = testInsert.Customers(name = "Custom Name")

    assertNotNull(row)
    assertEquals("Custom Name", row.name)
  }

  @Test
  def testDepartmentsInsert(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val row = testInsert.Departments()

    assertNotNull(row)
    assertNotNull(row.deptCode)
    assertNotNull(row.deptRegion)
    assertNotNull(row.deptName)
  }

  @Test
  def testProductsInsert(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val row = testInsert.Products()

    assertNotNull(row)
    assertNotNull(row.productId)
    assertNotNull(row.sku)
    assertNotNull(row.name)
    assertNotNull(row.price)
  }

  @Test
  def testAllScalarTypesInsert(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val row = testInsert.AllScalarTypes()

    assertNotNull(row)
    assertNotNull(row.id)
    assertNotNull(row.colNotNull)
  }

  @Test
  def testEmployeesWithDepartmentFK(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val dept = testInsert.Departments()

    val emp = testInsert.Employees(deptCode = dept.deptCode, deptRegion = dept.deptRegion)

    assertNotNull(emp)
    assertEquals(dept.deptCode, emp.deptCode)
    assertEquals(dept.deptRegion, emp.deptRegion)
  }

  @Test
  def testOrdersWithCustomerFK(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val customer = testInsert.Customers()

    val order = testInsert.Orders(customerId = customer.customerId.value)

    assertNotNull(order)
    assertEquals(customer.customerId.value, order.customerId)
  }

  @Test
  def testMultipleInserts(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val row1 = testInsert.Customers()
    val row2 = testInsert.Customers()
    val row3 = testInsert.Customers()

    assertNotEquals(row1.customerId, row2.customerId)
    assertNotEquals(row2.customerId, row3.customerId)
    assertNotEquals(row1.customerId, row3.customerId)
  }

  @Test
  def testInsertWithSeededRandom(): Unit = {
    val testInsert1 = TestInsert(Random(123))
    val testInsert2 = TestInsert(Random(123))

    val row1 = withConnection { c =>
      given java.sql.Connection = c
      testInsert1.Customers()
    }
    val row2 = withConnection { c =>
      given java.sql.Connection = c
      testInsert2.Customers()
    }

    assertEquals(row1.name, row2.name)
  }

  @Test
  def testDepartmentsGeneratesValidData(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val row = testInsert.Departments()

    assertNotNull(row)
    assertNotNull(row.deptCode)
    assertNotNull(row.deptRegion)
    assertNotNull(row.deptName)
  }
}
